From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Sun, 7 Nov 2077 00:00:00 +1400
Subject: [PATCH] Raytrace Entity Tracker

Original project: https://github.com/tr7zw/EntityCulling
Original license: Custom License

Original project: https://github.com/LogisticsCraft/OcclusionCulling
Original license: MIT

Ported by: https://github.com/LuminolMC/Luminol
Fixed and updated by: https://github.com/Winds-Studio/Leaf

diff --git a/net/minecraft/server/level/ChunkMap.java b/net/minecraft/server/level/ChunkMap.java
index 546b20f8998c71ca1a701de7efcedd8d821105e4..3585f941d3edd15276967ca09e83a635760d2cae 100644
--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -1281,7 +1281,7 @@ public class ChunkMap extends ChunkStorage implements ChunkHolder.PlayerProvider
                 double d1 = vec3_dx * vec3_dx + vec3_dz * vec3_dz; // Paper
                 double d2 = d * d;
                 // Paper start - Configurable entity tracking range by Y
-                boolean flag = d1 <= d2;
+                boolean flag = d1 <= d2 && !entity.isCulled(); // Leaf - Raytrace Entity Tracker
                 if (flag && level.paperConfig().entities.trackingRangeY.enabled) {
                     double rangeY = level.paperConfig().entities.trackingRangeY.get(this.entity, -1);
                     if (rangeY != -1) {
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 2786fbc43f238be058b153ba551f0a616c4a18f4..f69565586ef795daff48a1b75119db1d86d66b2e 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -133,7 +133,7 @@ import net.minecraft.world.scores.ScoreHolder;
 import net.minecraft.world.scores.Team;
 import org.slf4j.Logger;
 
-public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess, ScoreHolder, ca.spottedleaf.moonrise.patches.chunk_system.entity.ChunkSystemEntity, ca.spottedleaf.moonrise.patches.entity_tracker.EntityTrackerEntity {  // Paper - rewrite chunk system // Paper - optimise entity tracker
+public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess, ScoreHolder, ca.spottedleaf.moonrise.patches.chunk_system.entity.ChunkSystemEntity, ca.spottedleaf.moonrise.patches.entity_tracker.EntityTrackerEntity, dev.tr7zw.entityculling.versionless.access.Cullable {  // Paper - rewrite chunk system // Paper - optimise entity tracker // Leaf - Raytrace entity tracker
     public static javax.script.ScriptEngine scriptEngine = new javax.script.ScriptEngineManager().getEngineByName("rhino"); // Purpur - Configurable entity base attributes
     // CraftBukkit start
     private static final int CURRENT_LEVEL = 2;
@@ -5291,4 +5291,34 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         return this.distanceToSqr(entity.position());
     }
     // Leaf end - Optimize Entity distanceToSqr
+
+    // Leaf start - Raytrace entity tracker
+    private long lastTime = 0;
+    private volatile boolean culled = false;
+    private boolean outOfCamera = false;
+
+    @Override
+    public void setTimeout() {
+        this.lastTime = System.currentTimeMillis() + 1000;
+    }
+
+    @Override
+    public boolean isForcedVisible() {
+        return this.lastTime > System.currentTimeMillis();
+    }
+
+    @Override
+    public void setCulled(boolean value) {
+        this.culled = value;
+        if (!value) {
+            setTimeout();
+        }
+    }
+
+    @Override
+    public boolean isCulled() {
+        if (!org.dreeam.leaf.config.modules.misc.RaytraceTracker.enabled) return false;
+        return this.culled;
+    }
+    // Leaf end - Raytrace entity tracker
 }
diff --git a/net/minecraft/world/entity/EntityType.java b/net/minecraft/world/entity/EntityType.java
index adeab8980b590e4a8b64b62cb60f623b2a842982..67ec36f61184f1a7f2304ba860982cd12f0139a7 100644
--- a/net/minecraft/world/entity/EntityType.java
+++ b/net/minecraft/world/entity/EntityType.java
@@ -1064,6 +1064,7 @@ public class EntityType<T extends Entity> implements FeatureElement, EntityTypeT
     private final int clientTrackingRange;
     private final int updateInterval;
     public boolean dabEnabled = false; // Pufferfish
+    public boolean skipRaytraceCheck = false; // Leaf - Raytrace Entity Tracker
     private final String descriptionId;
     @Nullable
     private Component description;
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index 477455fdfcc591a89823e88983eb12dabb078d9b..24f6188f8996a41be5bbedbebabe392f99c5d41c 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -204,6 +204,11 @@ public abstract class Player extends LivingEntity {
     public int burpDelay = 0; // Purpur - Burp delay
     public boolean canPortalInstant = false; // Purpur - Add portal permission bypass
     public int sixRowEnderchestSlotCount = -1; // Purpur - Barrels and enderchests 6 rows
+    // Leaf start - Raytrace Entity Tracker
+    @Nullable private dev.tr7zw.entityculling.CullTask cullTask;
+    @Nullable private Level lastCullWorld;
+    private final boolean needToCull = !(this instanceof org.leavesmc.leaves.replay.ServerPhotographer);
+    // Leaf end - Raytrace Entity Tracker
 
     // CraftBukkit start
     public boolean fauxSleeping;
@@ -296,6 +301,37 @@ public abstract class Player extends LivingEntity {
         }
         // Purpur end - Burp delay
 
+        // Leaf start - Raytrace Entity Tracker
+        if (this.needToCull) {
+            if (!org.dreeam.leaf.config.modules.misc.RaytraceTracker.enabled) {
+                if (this.cullTask != null) {
+                    this.cullTask.signalStop();
+                    this.cullTask = null;
+                }
+            } else {
+                Level currLevel = this.level();
+                boolean needsUpdate = this.cullTask == null || this.lastCullWorld != currLevel;
+
+                if (needsUpdate) {
+                    if (this.cullTask != null) {
+                        this.cullTask.signalStop();
+                    }
+                    final com.logisticscraft.occlusionculling.OcclusionCullingInstance culling = new com.logisticscraft.occlusionculling.OcclusionCullingInstance(
+                        org.dreeam.leaf.config.modules.misc.RaytraceTracker.maxTraceDistance,
+                        new dev.tr7zw.entityculling.DefaultChunkDataProvider(currLevel)
+                    );
+                    this.cullTask = new dev.tr7zw.entityculling.CullTask(
+                        culling,
+                        this,
+                        org.dreeam.leaf.config.modules.misc.RaytraceTracker.boundingBoxLimit,
+                        org.dreeam.leaf.config.modules.misc.RaytraceTracker.traceInterval
+                    );
+                    this.cullTask.setup();
+                    this.lastCullWorld = currLevel;
+                }
+            }
+        }
+        // Leaf end - Raytrace Entity Tracker
         this.noPhysics = this.isSpectator();
         if (this.isSpectator() || this.isPassenger()) {
             this.setOnGround(false);
@@ -1553,6 +1589,12 @@ public abstract class Player extends LivingEntity {
         if (this.containerMenu != null && this.hasContainerOpen()) {
             this.doCloseContainer();
         }
+        // Leaf start - Raytrace Entity Tracker
+        if (this.cullTask != null) {
+            this.cullTask.signalStop();
+            this.cullTask = null;
+        }
+        // Leaf end - Raytrace Entity Tracker
     }
 
     public boolean isLocalPlayer() {
